UserMessageEvents Changes - Batch1

arprgr opened the pull request:
Completed
Create new email templates
	fromUserId, message not viewed
	toUserId, message waiting to be viewed
Add the below column to the message table
	state (0 default value message has not been viewed reminder may need to be sent if more than a hour, 2 message has not been viewed, reminder sent)
	(Changed Model and Migrations 20161209013101-update-messages  file)
	
Create a UserMessageEvents table to store message views
	eventId
	messageId
	clientTime
	toUserId
	fromUserId
	type
	createdAt
	updateAt
	new Model and Migration (create, update) Files

create a PUT messages_api api (createMessageViewedEvent) to be called on "user view message" event, this will insert a row in to UserMessageEvents table

Create a GET endpoint /api/events/user/:userid/message/:messageId that fetches all the events for a given user and message 


WIP
Create a batch to check if the user has viewed the message 
	for each message in messageTable 
		(message is a reminder, createdAt is more than one hour in the past and state equals zero) 
			search for given messageId in UserMessageEvents table
				if it does not exist
				send email to sender of the message (fromUserId)
				send email to receiver of the message 
				write an entry in the logfile
				state=2
				if it finds a record 
				state=3

Mocha Test cases for the above

arprgr committed:
Merge pull request #1 from EchNet/master

merging back fromEchnet

arprgr committed:
added state parameter

this will have the below values
const MESSAGE_STATE_UNCHECKED = 0;
const MESSAGE_STATE_UNREAD = 1;
const MESSAGE_STATE_REMINDER_SENT = 2;
const MESSAGE_STATE_VIEWED = 3;

arprgr committed:
User Message Event Changes

1. added State Parameter
 const MESSAGE_STATE_UNCHECKED = 0;
  const MESSAGE_STATE_UNREAD = 1;
  const MESSAGE_STATE_REMINDER_SENT = 2;
  const MESSAGE_STATE_VIEWED = 3;
2. added new UserMessageEvents table

arprgr committed:
UserMessageEvents table

needed to record user message action related events

arprgr committed:
udpate script for UserMessageEvents table

arprgr committed:
added createMessageViewedEvent api

arprgr committed:
email templates to be used for message events

arprgr committed:
dummy

arprgr committed:
added in error

arprgr committed:
dummmy

arprgr committed:
added i error

arprgr committed:
dummmy

arprgr committed:
new message events module

arprgr committed:
rolling back message events changes

these have been moved to api/events/user_message_events

arprgr committed:
new route: user message events

EchNet commented on server/migrations/20170321114511-update-usermessageevents.js:
```diff
@@ -0,0 +1,27 @@
+'use strict';
+module.exports = {
+  up: function(queryInterface, Sequelize) {
+    return queryInterface.addColumn('UserMessageEvents', 'fromUserId', {
```
fromUserId and toUserId are immutable properties of the Message.  They should not be replicated here in the UserMessageEvent.

EchNet replied:
Instead, we need a userId that identifies the *viewer* of the message.  This is usually the toUserId of the message, but not necessarily so.

EchNet commented on server/models/usermessageevents.js:
```diff
@@ -0,0 +1,177 @@
+
+  function associate(_models) {
+    models = _models;
+    UserMessageEvents.belongsTo(models.User, { as: "toUser" });
```
replace toUser and fromUser with simply user, as noted above.

EchNet commented on server/models/usermessageevents.js:
```diff
@@ -0,0 +1,177 @@
+    })
+ }
+
+ function findReadMessageEventsForSender(fromUserId) {
```
Remove this function.

EchNet commented on server/models/usermessageevents.js:
```diff
@@ -0,0 +1,177 @@
+    return UserMessageEvents.findAll(query);
+  }
+
+ function findReadMessageEventsForUser(toUserId) {
```
Good to have this but as noted previously, the user is the viewer of the message, not necessary the toUser or fromUser of the message.

EchNet commented on server/models/usermessageevents.js:
```diff
@@ -0,0 +1,177 @@
+        }
+        return this;
+      },
+      fromUser: function(fromUser) {
```
Just "user" as noted previously.

EchNet commented on templates/noticeMessageReceiver.pug:
```diff
@@ -0,0 +1,9 @@
+body
+  p Hello, #{message.toUserName}!
```
This should be message.toUser.name.

... assuming that message has been retrieved and joined with the user info.

EchNet commented on templates/noticeMessageReceiver.pug:
```diff
@@ -0,0 +1,9 @@
+body
+  p Hello, #{message.toUserName}!
+  p.
+    #{message.toUserName}! has sent you a message that has not been viewed by you. 
```
This should be message.fromUser.name

EchNet commented on templates/noticeMessageSender.pug:
```diff
@@ -0,0 +1,8 @@
+body
+  p Hello, #{message.fromUserName}!
```
message.fromUser.name

EchNet commented on templates/noticeMessageSender.pug:
```diff
@@ -0,0 +1,8 @@
+body
+  p Hello, #{message.fromUserName}!
+  p.
+    A message you have sent to #{message.toUserName}! has not been viewed. A similar notice is being sent out to them via email as well.
```
message.toUser.name

EchNet commented on server/api/messages_api.js:
```diff
@@ -1,152 +1,160 @@
-});
-
-module.exports = router;
+/* messages_api.js */
```
For some reason github is showing the entire contents of the two versions of the file side-by-side, instead of isolating the changes.  Any idea why?  This makes it difficult for me to find your changes.

EchNet commented on server/migrations/20161209013100-update-messages.js:
```diff
@@ -15,6 +15,12 @@ module.exports = {
         type: Sequelize.INTEGER
       });
     }).then(function() {
+      return queryInterface.addColumn('Messages', 'state', {
```
It's bad form to update existing migrations.  Please create a new migration.

EchNet commented on server/api/events/user_message_events.js:
```diff
@@ -0,0 +1,68 @@
+const Message = require("../../models/index").Message;
+
+const VALIDATOR = new ApiValidator({
+  state: {
```
User message event does not have a state field.  Going back to the spec, the fields are:
messageId  int, required
clientTime, string, required
type, string, default value is "open"

EchNet commented on server/api/events/user_message_events.js:
```diff
@@ -0,0 +1,68 @@
+  }
+});
+
+/* GET users listing. */
```
we don't need this GET method.

EchNet commented on server/api/events/user_message_events.js:
```diff
@@ -0,0 +1,68 @@
+});
+
+//Create Message Viewed Event
+router.put("/newMessageViewedEvent/:id", function(req, res) {
```
Create method should be a POST not a PUT

EchNet replied:
Path should be /api/events/user/:userId/message/:messageId as specified.

EchNet reviewed (changes requested):
Please make the changes specified and resubmit the pull request.

arprgr committed:
revisions based on Jim's comments

fromUserId and toUserId are immutable properties of the Message. They should not be replicated here in the UserMessageEvent.

arprgr committed:
revisions based on Jims comments

replace toUser and fromUser with simply user

arprgr committed:
revisions based on Jims Comments

arprgr committed:
revisions based on Jims comments

arprgr committed:
revisions based on Jim's comments

arprgr committed:
revisions based on Jim's comments

arprgr committed:
revisions based on Jim's comments

arprgr commented:
Hello Jim, all above comments have been addressed and relvant files with requisite changes have been checked into the arprgr repository.

arprgr committed:
message view reminders
